### PR TITLE
Add a fourth hint for `nix run`

### LINKED ISSUE #116340: help text for unfree outdated (and wrong)
```console
$ nix profile install nixpkgs#google-chrome
error: Package ‘google-chrome-89.0.4389.90’ in /nix/store/rmgzwxjxjhpk57cgzmxrvm6cqpa1h1ab-source/pkgs/applications/networking/browsers/google-chrome/default.nix:156 has an unfree license (‘unfree’), refusing to evaluate.

       a) To temporarily allow unfree packages, you can use an environment variable
          for a single invocation of the nix tools.

            $ export NIXPKGS_ALLOW_UNFREE=1

       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowUnfree = true; }
       in configuration.nix to override this.

       Alternatively you can configure a predicate to allow specific packages:
         { nixpkgs.config.allowUnfreePredicate = pkg: builtins.elem (lib.getName pkg) [
             "google-chrome"
           ];
         }

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allowUnfree = true; }
       to ~/.config/nixpkgs/config.nix.
(use '--show-trace' to show detailed location information)
```

```console
$ export NIXPKGS_ALLOW_UNFREE=1
$ bat ~/.config/nixpkgs/config.nix
───────┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: /home/blaggacao/.config/nixpkgs/config.nix
───────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ { allowUnfree = true; }
   2   │
───────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

```console
$ nix profile install nixpkgs#google-chrome
error: Package ‘google-chrome-89.0.4389.90’ in /nix/store/rmgzwxjxjhpk57cgzmxrvm6cqpa1h1ab-source/pkgs/applications/networking/browsers/google-chrome/default.nix:156 has an unfree license (‘unfree’), refusing to evaluate.

       a) To temporarily allow unfree packages, you can use an environment variable
          for a single invocation of the nix tools.

            $ export NIXPKGS_ALLOW_UNFREE=1

       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowUnfree = true; }
       in configuration.nix to override this.

       Alternatively you can configure a predicate to allow specific packages:
         { nixpkgs.config.allowUnfreePredicate = pkg: builtins.elem (lib.getName pkg) [
             "google-chrome"
           ];
         }

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allowUnfree = true; }
       to ~/.config/nixpkgs/config.nix.
(use '--show-trace' to show detailed location information)
```

at least `a)` & `c)` are misleading / wrong.

```console
$ nix --version
nix (Nix) 2.4pre20210308_1c0e3e4
```

### PR DESCRIPTION
**Pull request by waxlamp**
###### Motivation for this change

Attempting to run `postman` via `nix run nixpkgs#postman`, I got an [error message](https://github.com/NixOS/nixpkgs/blob/d600f006643e074c2ef1d72e462e218b647a096c/pkgs/stdenv/generic/check-meta.nix#L122-L136) about unfree packages, but the mitigations listed weren't helping. With some discussion from [this discourse topic](https://discourse.nixos.org/t/nix-run-does-not-install-unfree-packages-nixpkgs-allow-unfree-doesnt-seem-to-help/12039), I understood that I needed the `--impure flag`, but the error message does not mention that.

This PR just adds a bit of text to that error message block about instituting one of the recommended fixes, then making sure to use `--impure` with `nix run`.

I am not sure how to test this change; some help with that would be appreciated.

Closes #116340.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/stdenv/generic/check-meta.nix [hunk 1]
```diff
@@ -133,6 +133,9 @@ let
       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allow${allow_attr} = true; }
       to ~/.config/nixpkgs/config.nix.
+
+      d) If you are using `nix run`, in addition to one of the fixes above, you
+      may need to use the `--impure` flag as well.
```
Comment (Pacman99) [inline pkgs/stdenv/generic/check-meta.nix:138]:
```suggestion
      d) If you are using nix unstable commands, in addition to one of the fixes above, you
      may need to use the `--impure` flag as well.
```
Not sure the best way to word that, but this applies to build, run, shell and other unstable commands.

Comment (SuperSandro2000) [reply pkgs/stdenv/generic/check-meta.nix:138]: Is that part of nix-command or flakes experimental feature?

Comment (Pacman99) [reply pkgs/stdenv/generic/check-meta.nix:138]: I'm not sure, one of them makes all evaluation pure by default. I think its flakes experimental feature. But usually you enable both, so I don't know.

Comment (waxlamp) [reply pkgs/stdenv/generic/check-meta.nix]: Sorry for the delay, @Pacman99, and thank you for the review.

I tried to make the information more comprehensive (ade15a5), but I also don't know the answer to @SuperSandro2000's question, so am not totally sure how best to phrase it. Please take a look at my update and let me know what you think.

Thanks!

**Review by Pacman99 — APPROVED** [review]
I like the new wording, this looks good. Also this closes #116340, so it would be good to link that in the PR.

**Comment by waxlamp** [comment]
@Pacman99, what else should I do to get this ready for merge?

File: pkgs/stdenv/generic/check-meta.nix (hunk 1 above)
Comment (NobbZ) [reply pkgs/stdenv/generic/check-meta.nix]: I consider the examples problematic, `run` and `build` exist for 2.3 as well, and `shell` also IIRC.

Can't you say "in a pre release version of nix"?

Comment (waxlamp) [reply pkgs/stdenv/generic/check-meta.nix]: @nobbz, could you maybe provide some example wording? I'm not sure how to update the text to resolve your concerns.

File: pkgs/stdenv/generic/check-meta.nix [hunk 2]
```diff
@@ -133,6 +133,10 @@ let
       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allow${allow_attr} = true; }
       to ~/.config/nixpkgs/config.nix.
+
+      d) If you are using the unstable `nix` commands (e.g., `nix run`, `nix
+      build`, `nix shell`, etc.), in addition to one of the fixes above, you
+      may need to use the `--impure` flag as well.
```
Comment (jonringer) [inline pkgs/stdenv/generic/check-meta.nix:139]: Personally, I would add a little "why" this is necessary. Also word it in a way to hint that user local overlays and other settings will also be "unleashed" when using `--impure`
```suggestion
      d) If you are using the unstable `nix` commands (e.g., `nix run`, `nix
      build`, `nix shell`, etc.), then nix will evaluate in "restrict-eval" mode which
      denies it the ability to read a user's local configuration. Add `--impure`
      to enable reading from the user's nix config directory.
```

Comment (waxlamp) [reply pkgs/stdenv/generic/check-meta.nix]: Thanks for the suggestion, @jonringer!

**Comment by colemickens** [comment]
I don't like this change. I'd much rather see this warning instruct people on how to do the right thing with pure-eval, rather than just encouraging more indirect overriding of internal behavior.

File: pkgs/stdenv/generic/check-meta.nix (hunk 1 above)
Comment (NobbZ) [reply pkgs/stdenv/generic/check-meta.nix]: "If you are using a prerelease version of nix (eg. flake enabled), you need to use the `--impure` flag in combination with solution a) or c)"

**Comment by waxlamp** [comment]
> I don't like this change. I'd much rather see this warning instruct people on how to do the right thing with pure-eval, rather than just encouraging more indirect overriding of internal behavior.
> 
> I can see it now - someone will do this and then wind up including more impurity than they'd wanted to and then being confused why allowing unfree with `--impure` is somehow causing overlays to be loaded, etc.

I'm getting a lot of conflicting advice on this PR and would like some help in knowing which direction is best for everyone. For my part, I found the existing help text to be both unhelpful and, after diving into things, inaccurate, so I was trying to improve both aspects in the way that minimally made me able to function.

I'm not aware in general of how to "do the right thing with pure-eval"; would you mind pointing me towards some resources that explain it? Related request: would you mind prompting me towards what sort of help text you think would be better?

If we could form some consensus on this, I think that would be great. As it stands, I don't have enough knowledge to understand how best to improve the error message here.

Thanks everyone!

**Comment by NobbZ** [comment]
The "right" thing, is often considered a wrapping flake, that imports `nixpkgs` with an appropriate config and re-exposes all of `nixpkgs` package set.

This is something I do not like though, as it requires a lot of work, especially if you want to have some certain commit of `nixpkgs` or use different branches all the time…

`--impure` + env-vars is in my opinion usually the easiest and quickest way to test something broken/unfree without having to enable the setting in config/rebuild/switch/rollback on misbehave…

**Comment by waxlamp** [comment]
> The "right" thing, is often considered a wrapping flake, that imports `nixpkgs` with an appropriate config and re-exposes all of `nixpkgs` package set.
> 
> This is something I do not like though, as it requires a lot of work, especially if you want to have some certain commit of `nixpkgs` or use different branches all the time…
> 
> `--impure` + env-vars is in my opinion usually the easiest and quickest way to test something broken/unfree without having to enable the setting in config/rebuild/switch/rollback on misbehave…

This confirms my own intuition: the "right thing" is some deeper action that is bulky to perform and requires a lot of understanding of theory; the thing that lets me continue being productive in my environment immediately is using the `--impure` flag.

**Comment by waxlamp** [comment]
@jonringer, thanks again for your input on the help text. I am not sure what else to do to get this reviewed and merged. Can you help push this along?

**Comment by SuperSandro2000** [comment]
Similar thing was already done in 7bea56b4253a3232a44249b3dd7d2a82795ef7fc

**Pull request closed without merging**